lib/blocks.php

```diff
@@ -216,3 +216,189 @@ function gutenberg_register_legacy_social_link_blocks() {
 	}
 }
 add_action( 'init', 'gutenberg_register_legacy_social_link_blocks' );
+
+/**
+ * Renders the classNames and styles for blocks
+ *
+ * @param string $block_content Output of the current block.
+ * @param array  $block Block Object.
+ * @return string New block output.
+ */
+function gutenberg_experimental_apply_classnames_and_styles( $block_content, $block ) {
+	if ( ! isset( $block['attrs'] ) ) {
+		return $block_content;
+	}
+
+	$block_type = WP_Block_Type_Registry::get_instance()->get_registered( $block['blockName'] );
+	// If no render_callback, assume styles have been previously handled.
+	if ( ! $block_type || ! $block_type->render_callback ) {
+		return $block_content;
+	}
+	// Check what style features the block supports.
+	$supports = gutenberg_experimental_global_styles_get_supported_styles( $block_type->supports );
+
+	$attributes = array();
+	$attributes = gutenberg_experimental_build_css_colors( $attributes, $block['attrs'], $supports );
+	$attributes = gutenberg_experimental_build_css_typography( $attributes, $block['attrs'], $supports );
+
+	if ( ! count( $attributes ) ) {
+		return $block_content;
+	}
+
+	$dom = new DOMDocument( '1.0', 'utf-8' );
+
+	// Suppress warnings from this method from polluting the front-end.
+	// @codingStandardsIgnoreStart
+	if ( ! @$dom->loadHTML( $block_content, LIBXML_HTML_NOIMPLIED | LIBXML_HTML_NODEFDTD | LIBXML_COMPACT ) ) {
+	// @codingStandardsIgnoreEnd
+		return $block_content;
+	}
+
+	$xpath      = new DOMXPath( $dom );
+	$block_root = $xpath->query( '/*' )[0];
+
+	if ( empty( $block_root ) ) {
+		return $block_content;
+	}
+
+	// Some inline styles may be added without ending ';', add this if necessary.
+	$current_styles = trim( $block_root->getAttribute( 'style' ), ' ' );
+	if ( strlen( $current_styles ) > 0 && substr( $current_styles, -1 ) !== ';' ) {
+		$current_styles = $current_styles . ';';
+	};
+
+	// Merge and dedupe new and existing classes and styles.
+	$classes_to_add = esc_attr( implode( ' ', array_key_exists( 'css_classes', $attributes ) ? $attributes['css_classes'] : array() ) );
+	$styles_to_add  = esc_attr( implode( ' ', array_key_exists( 'inline_styles', $attributes ) ? $attributes['inline_styles'] : array() ) );
+	$new_classes    = implode( ' ', array_unique( explode( ' ', ltrim( $block_root->getAttribute( 'class' ) . ' ' ) . $classes_to_add ) ) );
+	$new_styles     = implode( ' ', array_unique( explode( ' ', $current_styles . ' ' . $styles_to_add ) ) );
```

Comment:
If I am reading this correctly, the separator for `implode` and `explode` should be `';'` instead of `' '`, otherwise `array_unique` doesn't have much to trim.

Same for two lines above at the assignment of `$styles_to_add`.

Using a `print_r` and a `die` on my env:

```
INPUT:
background-color:#b44343;min-height:244px;

OUTPUT:
Array
(
    [0] => background-color:#b44343;min-height:244px;
    [1] => 
)
```

Reply:
Do the styles in `$current_styles` never have spaces between them like in your example input?  or could they be either case?  

It does look like we need to update these so that they are more consistent with each other and can properly remove the duplicates.  We can implode `$styles_to_add` with no spacing (since `';'` is already added to the values in the array), and explode/implode `$new_styles` on `';'` if there are never spaces.  If there can be I guess we need an extra step to trim whitespace for the `array_unique` to compare properly? 🤔

Reply:
> Do the styles in `$current_styles` never have spaces between them like in your example input? or could they be either case?

I didn't check to see if this was guaranteeable, but those are the styles I got after using the actual user-facing tools. I'd look at the framework to make sure. If it seems like spaces could make their way in, you can always use `preg_split`:

```php
php > var_dump( preg_split( '/; ?/', 'color:red;color:green; color:blue' ) );
array(3) {
  [0]=>
  string(9) "color:red"
  [1]=>
  string(11) "color:green"
  [2]=>
  string(10) "color:blue"
}
```

Of course, you can take it as far as you like:

```php
php > var_dump( preg_split( '/ *; */', 'color:red   ; color:green; color:blue' ) );
array(3) {
  [0]=>
  string(9) "color:red"
  [1]=>
  string(11) "color:green"
  [2]=>
  string(10) "color:blue"
}
```

Reply:
> We can implode `$styles_to_add` with no spacing (since `';'` is already added to the values in the array)

I'm not sure what you mean here. When wouldn't `explode`/`preg_split` strip semi-colons from the captured styles?

Reply:
Ah yeah `preg_split` could be the way to go then.

> I'm not sure what you mean here. When wouldn't explode/preg_split strip semi-colons from the captured styles?

I just mean that currently the values added to `$attributes['inline-styles']` all currently have the semicolons at the end of the string already.  So that array looks something like `[ '--wp--style--color--link: #fff;' , 'background: #000;' ]` etc.  So when `implode` happens on the line with `$styles_to_add` that turns into `'--wp--style--color--link: #fff; background: #000;'` etc, even though we are just imploding with `' '` as the separator.

Reply:
Thanks for pointing this out by the way.  I hope to address this relatively shortly but have been preoccupied so far today.

Reply:
I proposed a fix in #24486